src/Microsoft.OData.Core/RefOfT.cs

```diff
@@ -0,0 +1,60 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="RefOfT.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData
+{
+    /// <summary>
+    /// Asynchronous methods do not support ref and out parameters.
+    /// This class provides a workaround to enable passing around
+    /// of the value in an object.
+    /// </summary>
+    /// <typeparam name="T">The type of object wrapped in the object.</typeparam>
+    internal sealed class Ref<T>
+    {
+        private T value;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Ref{T}"/> class.
+        /// </summary>
+        public Ref() : this(default(T)) {}
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Ref{T}"/> class with value to be wrapped.
+        /// </summary>
+        /// <param name="value">The value to be wrapped.</param>
+        public Ref(T value)
+        {
+            this.value = value;
+        }
+
+        /// <summary>
+        /// Gets or sets the value wrapped in the object.
+        /// </summary>
+        public T Value
+        {
+            get { return this.value; }
+            set { this.value = value; }

```

Comment:
>set { this.value = value; } [](start = 12, length = 27)

If we have the constructor to set the value, shall we keep the public set for the property?

Reply:
@xuzhg Keeping the property setter public allows us do something like [this](https://github.com/OData/odata.net/pull/2002/files#diff-44795ee4fde32bffe0ac82d8b21b2ac93180517dce7d66dfeb7afce151a1fb8bR408)
```csharp
buffer.Value = BufferUtils.InitializeBufferIfRequired(bufferPool, buffer.Value);
```